### ISSUE TITLE
Custom JWT backend with CredentialsProvider JWT

### ISSUE THREAD

**Issue by Skyquek** [nextauthjs/next-auth#8086]
### Question 💬

I am using Django with strawberry Django Auth https://github.com/nrbnlulu/strawberry-django-auth where its use JWT token to authenticate user. In my nextjs app, I am using https://next-auth.js.org/providers/credentials for the authentication library. However, I realize that the `authorization` function actually wrap the user data object into its own JWT. 

So, this would mean I need to wrap user token to the user object as well? Or I should change it to my backend JWT? 

### How to reproduce ☕️

This is my [...nextauth].ts.

```typescript
import NextAuth, { NextAuthOptions } from "next-auth";
import CredentialsProvider  from "next-auth/providers/credentials";
import client from "../../../apollo-client";
import { gql } from "@apollo/client";
import { getLogger } from "../../../logging/log-util";

// For more information on each option (and a full list of options) go to
// https://next-auth.js.org/configuration/options
export const authOptions: NextAuthOptions = {

  // https://next-auth.js.org/configuration/providers/oauth
  providers: [
    CredentialsProvider({
      // How to delete this? I have my own signin page.
      name: "Credentials (Deprecated)",
      credentials: {
        username: { label: "Username", type: "text", placeholder: "jsmith" },
        password: { label: "Password", type: "password" }
      },
      async authorize(credentials) {
        // Add logic here to look up the user from the credentials supplied        
        if (credentials?.username === undefined || credentials?.password === undefined) { 
          return null;
        } else {
          let username = credentials.username;
          let password = credentials.password;

          // let username = "new_user4";
          // let password = "SuperSecureP@sw0rd";

          const { data } = await client.mutate({
            mutation: gql`
            mutation {
              tokenAuth(username: "${username}", password: "${password}") {
                success
                errors
                refreshToken {token created revoked expiresAt isExpired __typename}
                token {
                  payload {
                    origIat
                    exp
                  }
                  token
                }
                user {
                  isActive
                  username
                  email
                  status {
                    verified
                  }
                }
              }
            }  
            `,
          });

          // console.log(data.tokenAuth.success);
          if (data.tokenAuth.success === true) {
            // Any object returned will be saved in `user` property of the JWT
            const user = {
              "serverRefreshToken": data.tokenAuth.refreshToken.token,
              "isActive": data.tokenAuth.user.isActive,
              "username": data.tokenAuth.user.username,
              "email": data.tokenAuth.user.email
            }
            return user;
          } else {
            // console.log('error lel');
            // console.error(data);
             // If you return null then an error will be displayed advising the user to check their details.
            //  return Error('Could not log you in.');
            return null;
             // You can also Reject this callback with an Error thus the user will be sent to the error page with the error message as a query parameter
          }
        }
      }
    })
  ],
  // callbacks: {
  //   async jwt({ token }) {
  //     console.log("jwt callback");
  //     console.log(token.email);
  //     return token
  //   },
  // },
}

export default NextAuth(authOptions)
```

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by arminhupka** [comment]
If you want make call with your api token you must put this token into JWT by jwt callback. I see that you want refresh token but unfortunately it's not working so far in next-auth. You can make request for new token but you won't be able to update already stored token in next-auth token.

**Comment by namdien177** [comment]
>  I see that you want refresh token but unfortunately it's not working so far in next-auth. You can make request for new token but you won't be able to update already stored token in next-auth token.

I think we need to clarify as this is [an ongoing bug of nextjs and next-auth](https://github.com/nextauthjs/next-auth/discussions/6642)

For your question, I also have to spend a lot of time just fiddling around the JWT callback for different providers so I will explain what I understood (since the doc seems to be pretty lacking and the example is pretty outdated)

> So, this would mean I need to wrap user token to the user object as well?
Yes, you should. Or you can just return the accessToken and decode it in the `jwt` callback. The flow should be like this:

Authorize returns `data` => that data will be available *once* (at the moment login only) at the `jwt` callback.
`jwt` callback return `session data` => that data will always be reflected in the `session` callback.

```ts
/**
   * The returned value will be encrypted, and it is stored in a cookie.
   *
   * @param account Only passed the first time this callback is called on a new session
   * @param token Always available with what the OAuth provider returns (at first time) or from the second time onward.
   *              Note: for credentials, this object is present, but all information is `null`
   *                    - not sure if it's the correct behavior since I'm using a custom adapter.
   * @param user Only passed the first time this callback is called on a new session
   * @param profile Only passed the first time this callback is called on a new session
   * @param isNewUser Only passed the first time this callback is called on a new session
   */
async jwt({ account, token, user, profile, isNewUser }) {
    const data = { ...token };

    if (user && user.accessToken) {
      // If login with Credentials
	  // - user object is what the credential provider returns (from the authorize function)
      // - for instance, you return `accessToken` in the authorize function => bind accessToken to the data
      // and make it available to the client.
      data.sub = user.id;
      data.email = user.email;
      data.name = user.name;
      data.image = user.image;
	  //You can bind more information as I did above for convenience.
      data.accessToken = user.accessToken;
      return data;
    }

    if (user && user.id) {
      // If login with OAuth (GitHub, Google, etc)
      // - This is what the `user` of the provider (stored in your database) returns.
      // - Normally you will want to request accessToken from your custom backend,.
      // - bind accessToken to `data`
      const {
        data: { accessToken },
      } = await askAuthenticatedViaProviders( . . . );
      // forward the token to the client, just like the credential provider above
      data.accessToken = accessToken;
      return data;
    }

    /**
     * After the 2nd call.
     * Job you might want to do:
     * - data.accessToken exists from the second call (we bind it above)
     * - check expire time of the token.
     * - Refresh if necessary
     * - return the value (will be passed to Session callback)
     *   -> Should not return sensitive data!
     * - Currently broken due to bug https://github.com/nextauthjs/next-auth/discussions/6642
     */
    const payload = jwt.decode(data.accessToken, { json: true });
    if (!payload) {
      console.error(chalk.red('AccessToken after 2nd call is invalid:', data));
      throw new Error('Unable to decode');
    }
    const { exp, id } = payload;

    if (exp && id) {
      const timeEpoch = exp * 1000; // seconds to milliseconds
      if (dayjs(timeEpoch).isBefore(new Date())) {
        // refresh
        console.log(chalk.yellow('should refresh token'));
        const {
          data: { accessToken },
        } = await refreshAccessToken( . . . );
        data.accessToken = accessToken; // currently not work...
      }
    }

    return data;
  },

  async session({ session, user, token, newSession, trigger }) {
    const alteredSession = { ...session };
    alteredSession.user.id = token.sub;
    // This will be returned to the client so be careful what to send here.
    return {
      ...alteredSession,
      accessToken: token.accessToken,
    };
  },
```

**Comment by rexfordessilfie** [comment]
I implemented something similar in the past, but between a Next.js application and an Express server. This was the scenario:
1. I have an Express backend, separately hosted
2. I have a Next.js application, using NextAuth, and `CredentialsProvider` that allows users to authenticate with a user name and password
3. After a user authenticates in the Next.js application, they can access secured/protected routes in the Express backend

I only completed this solution locally, but then switched to using Next.js as the backend solution before deploying, hence I did not execute this solution all the way to completion. I am still sharing where I got to (and additional thoughts here) just in case it helps you. Also ideas from the previous answers go into more detail for certain pieces (with respect to refresh tokens which I have not implemented).

## Solution
Here's a high-level description of everything that I remember doing for this to work. Assuming everything is in-place with CORS and other cross-site request concerns, you want the following:
1. Both the Next.js app and backend use the same: JWT secret, JWT algorithm to encode and decode JWT payloads
2. Backend knows where to find, decode and check the JWT token as set by the app
3. This is not necessarily a requirement, but if the backend requires a strict shape/fields for the contents of the JWT payload, then the Next.js app must ensure these fields are present in the JWT.

## Approach
Here's a minimal approach to ensuring all of these conditions.

### 1. Using the same secret + algorithm to encode/decode JWT 
Here `createJWT` and `decodeJWT` are your own implementations of JWT encoding/decoding that use the same JWT algorithm(s) as used by your backend. In my case, I used functionality from `jsonwebtoken` npm package. Also here the `secret` passed is the same as in your `NEXTAUTH_SECRET` from your environment variables. You can also ignore the `secret` provided in the callback and use one you have defined elsewhere.

```ts
export const authOptions: NextAuthOptions = {
  ...,
  jwt: {
    encode: async ({ token, secret }) => {
      if (token && secret) {
        return createJWT(token, secret);
      }
      return '';
    },
    decode: async ({ token, secret }) => {
      if (token && secret) {
        return decodeJWT(token, secret);
      }
      return null;
    },
  },
 ...,
}
```

### 2. Backend knows where to find, decode and check the JWT token as set by the app
For your backend to find and decode the JWT token created by your Next.js app, it would have to access the JWT cookie from the request payload. NextAuth uses either `"next-auth.session-token"` or `"__Secure-next-auth.session-token"`. [source](https://github.com/nextauthjs/next-auth/blob/48b9a0203e7881d4d071e0a5d4bd7a9d1bbe5184/packages/next-auth/src/jwt/index.ts#L77)

Extracting the cookie from the request would look different for different backend frameworks (esp. Django which I am not very familiar with), so I'll leave that to you!

> NB: you would have to have properly setup CORS, and to include credentials when making requests from Next.js app land to your backend for this to work in the first place.

### 3. JWT Token Contents
On top of standard JWT fields, some backends may also require the presence of certain fields in the payload. If your backend has this requirement, your Next.js app can ensure such fields are present in a two step process. First return them in in the `authorize` callback of `CredentialsProvider` so they are returned as the `user` object in the `jwt` callback in `authOptions.callbacks.jwt`.

**Part A: returning necessary fields in `CredentialsProvider`, `authorize` callback.**
Whatever is returned in the `authorize` callback here is passed back into the `jwt` callback in `authOptions.callbacks`.

```ts
export const authOptions: NextAuthOptions = {
  ....,
  providers: [
    CredentialsProvider({
      id: 'credentials',
      ...,
      async authorize(credentials) {
        if (!credentials?.identifier || !credentials.password) {
          throw new Error('Missing credentials');
        }
        
        // Authenticate the user (you are already doing this with graphql but I redacted that to a simple function to keep things simple)
        const result = authenticateUser(credentials);

        // Return details to be provided back into the `user` property of jwt. (You are doing this already again!)
        return { id: result.id, fieldA: result.fieldA, ... }
      },
    }),
  ],
  ...,
};
```


**Part B: placing right information into JWT on Next.js app side using the `jwt` callback in `authOptions.callbacks`**
Anything returned by this callback is encoded as part of the JWT. Here we ensure that it has the right details, once we determine that we are dealing with a `'credentials'` account type.

```ts
export const authOptions: NextAuthOptions = {
  ...,
  callbacks: {
    async jwt({ account, token, user }) {
      if (account?.type === 'credentials') { // OR any check to ensure you are working with 'credentials' provider :)
        token.id = user?.id;
        token.fieldA = user.fieldA; // Copy over required JWT fields into the token object
      }

      return token;
    },
  },
  ...,
};
```

### Typescript
You will probably run into some type issues using Typescript. You can fix these by [augmenting next-auth's types](https://next-auth.js.org/getting-started/typescript#module-augmentation) to allow custom fields such as `fieldA` on the `user` and `token` objects.

That's it. Let me know if there is any more detail missing/that you require for your specific problem. Thanks.

**Comment by namdien177** [comment]
> Is there a need to saved the token into the cookies or I can just saved in the user object? I am worry later in the implementation, will have issue where I need to access protected route?

In my case, I return the `accessToken` together with the user object to the client via session callback. NextAuth should handle the storing for you (in cookies) so you won't need to care about that.

Since you will have `accessToken` available right on the client (using `useSession` will get you the `data` which is what you return from session callback), you can easily attach it to the header of any API request which goes directly to your external backend.

```typescript
const { data } = useSession();

const headers = new Headers();
if (data.accessToken) {
  headers.set('Authorization`, `Bearer ${data.accessToken}`);
}

// then consume the headers in your API request.
const response = await ky.get(`/url`, { headers }).json();

```

**Comment by namdien177** [comment]
The answer from @rexfordessilfie awesome and detailed. Correct me if I'm wrong, but that solution uses the token from the cookies, which I believe will not work if you are using cross-domain ([even with sub-domain in some cases](https://stackoverflow.com/questions/59785832/samesite-cookies-frames-sub-domains-and-redirections)) with secured methods (POST) for example [even if its default setting for `sameSite` is `lax`](https://next-auth.js.org/configuration/options#cookies). I haven't tried this yet so I might get it wrong.

Most tutorials for using external servers will suggest having different `accessToken` returns from the external backend. This is not optimal as the ideal solution of @rexfordessilfie , since it will have a double encode/decode process and will require you to have 2 separate authentication tokens; but it's more simple and guarantees to work cross-domain since you will have easy access to your external backend token with `useSession` on the client to make the request directly.

> Is there a need to saved the token into the cookies or I can just saved in the user object?

I believe what you return from `session` callback will be automatically handled by this library so you won't need to care much about where to store or how to get the token to authenticate the user. Although, updating the data (refresh token) will be impossible for now with the bug I mentioned above.

**Comment by rexfordessilfie** [comment]
Yep @namdien177. Spot on. My answer is most suitable for cookie-based session storage + authentication, and also gets tricky with cross-site requests. 

If you will be interacting with the backend using Authorization Headers, then it would suffice to have the token returned from the custom backend returned in the `jwt` callback and then also in the `session` callback to have it available to the client. Essentially encoding a JWT within a JWT, it sounds like?

This way you do not need to ensure that both the Next.js app and the backend utilize the same JWT secret + token strategy. 

The solution I shared worked fine for my (local) use-case and for cookie-based authentication. I did not expound on the CORS details, but a part of it also involved setting up a [proxy](https://nextjs.org/docs/pages/api-reference/next-config-js/rewrites). There's an article on how to implement an actual API proxy [here](https://maxschmitt.me/posts/next-js-api-proxy), but it is more involving!
